src/adapter/src/coord.rs

```diff
@@ -1375,13 +1386,21 @@ pub struct Coordinator {
     /// Data used by the statement logging feature.
     statement_logging: StatementLogging,
 
-    /// Limit for how many conncurrent webhook requests we allow.
+    /// Limit for how many concurrent webhook requests we allow.
     webhook_concurrency_limit: WebhookConcurrencyLimiter,
 
     /// Optional config for the Postgres-backed timestamp oracle. This is
     /// _required_ when `postgres` is configured using the `timestamp_oracle`
     /// system variable.
     pg_timestamp_oracle_config: Option<PostgresTimestampOracleConfig>,
+
+    /// Periodically asks cluster scheduling policies to make their decisions.
+    check_scheduling_policies_interval: tokio::time::Interval,
+
+    /// This keeps the last On/Off decision for each cluster and each scheduling policy.
+    /// (Clusters that have been dropped or are otherwise out of scope for automatic scheduling are
+    /// periodically cleaned up from this Map.)
+    scheduling_decisions: BTreeMap<ClusterId, BTreeMap<&'static str, bool>>,
```

Comment:
Why do we need this map? To me it seems like `async fn check_scheduling_policies(...)` has enough info, i.e. the Catalog and Compute Controller, to know whether or not replicas should be started or stopped. My concern is this map somehow getting out of sync with the rest of our state and replicas not getting scheduled as they should

Reply:
This is because of the "certain future policies might be called not just together with all the other policies" part of the PR description. If some policies will run in isolation (i.e., not together with other policies), then we have to have the latest decisions of all other policies available to be able to act on the decision of the policy that we ran in isolation.

Reply:
Ahhh I think I understand! More concretely would a policy that runs in isolation be something like `PAUSE CLUSTER foo`? That might come from a different code path and when we decide whether or not to actually drop the replicas we need to know what caused the replica to turn on?

If that's the case, this kind of feels like something that should be persisted in the Catalog state? For example, if a user paused a cluster and then `environmentd` restarted because of an upgrade we would want to remember that policy decision and keep the replica paused on restart. Definitely don't need to do this for the current version, but I think this would also solve the problem of losing the user's desired `replication_factor` since the "status/state" of the cluster could be represented separately.

Reply:
> More concretely would a policy that runs in isolation be something like PAUSE CLUSTER foo?

With the current architecture, such a policy is not really possible to implement. One reason is that the `scheduling_decisions` map is not persisted in the catalog (as you also pointed out). Another reason is that `handle_scheduling_decisions` just ORs all the On decisions. So even if the PAUSE CLUSTER policy says Off, other policies might still say On, and then the cluster wouldn't be turned off.

A different example for a policy that might be called in isolation would be a policy for turning on clusters in response to peeks. We probably want to call such a policy from the peek handling code, where other policies are not called. So, how I imagine implementing such a policy in more detail is that it would base its On/Off decision on whether there is an outstanding peek (or had one in the last N minutes). This way, it's also ok to call it periodically in `check_scheduling_policies`, but the reason for calling it in the peek handling code is so that we can react more quickly to a peek.

Reply:
Btw. if the user want to pause an auto-scheduled cluster, she can just `ALTER CLUSTER SET (SCHEDULE = MANUAL)`, and then change the replication factor to 0.